[CELEBORN-696] Fix bugs related with shutting down and excluded workers

waitinfuture opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  - Make sure the PR title start w/ a JIRA ticket, e.g. '[CELEBORN-XXXX] Your PR title ...'.
  - Be sure to keep the PR description updated to reflect all changes.
  - Please write your PR title to summarize what this PR proposes.
  - If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
1. Foreach PartitionLocation returned from Register Shuffle, remove from worker's local excluded list to refresh the local information.
2. ChangeLocationResponse will also return whether oldPartition is excluded in LifecycleManager. If so, remove it from Executor's excluded list.
3. Always trigger commit files for shutting down workers returned from HeartbeatFromApplicationResponse.
4. HeartbeatFromWorker sends the correct disk infos regardless of shutting down or not.

After this PR, the priority of excluded list is Master > LifecycleManager > Executor.

### Why are the changes needed?
During test with graceful turned on(workers have static rpc/push/fetch/replicate ports) and consistently restart one out of three workers, I encountered several bugs.
1. First I killed worker A, then Executor's client's local excluded list will contain A, after A stopped, I started it again, then master will offer slots on A, so we should remove from the executor's excluded list then.
2. When I kill-and-start a worker twice in a short time smaller than the app heartbeat interval, the second time WorkerStatusTracker will not trigger commit files because the local cache for the worker has not been refreshed.
3. When a worker is shutting down, in its heartbeat it passes empty diskInfos, and master blindly added to excluded list. We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.


### Does this PR introduce _any_ user-facing change?
Yes, it fixes several bugs described above.


### How was this patch tested?
Manual test.

waitinfuture commented:
cc @RexXiong @AngersZhuuuu @pan3793

RexXiong commented on client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:
```diff
@@ -198,6 +201,6 @@ class WorkerStatusTracker(
     val shutdownList = newShutdownWorkers.asScala.filterNot(shuttingWorkers.asScala.contains).asJava
     shuttingWorkers.addAll(newShutdownWorkers)
-    shutdownList
+    newShutdownWorkers
```
If so, we can simplify this code.

zhongqiangczq replied:
This function's origin meaning is that only newly added shutdownworkers can be returned (which is not in shuttingWorkers ), but now it returns all the shutdownworkers. Flink's shufflemaster will register ShuffleResourceTracker as workerStatusListeners. ShuffleResourceTracker should juse processes the newly add shutdownworkers. Is that right ? @RexXiong

RexXiong replied:
ShuffleResourceTracker of Flink doesn't care shutdown workers currently @zhongqiangczq

RexXiong commented on master/src/main/java/org/apache/celeborn/service/deploy/master/clustermeta/AbstractMetaManager.java:
```diff
@@ -209,7 +209,7 @@ public void updateWorkerHeartbeatMeta(
     }
     appDiskUsageMetric.update(estimatedAppDiskUsage);
-    if (!blacklist.contains(worker) && disks.isEmpty()) {
+    if (!blacklist.contains(worker) && disks.isEmpty() && !shutdownWorkers.contains(worker)) {
```
> Right, so I think it's better to alway send the true status regardless it's shutting down or not, WDYT?
> 
> ```
>     val diskInfos =
>       if (shutdown.get()) {
>         Seq.empty[DiskInfo]
>       } else {
>         storageManager.updateDiskInfos()
>         workerInfo.updateThenGetDiskInfos(storageManager.disksSnapshot().map { disk =>
>           disk.mountPoint -> disk
>         }.toMap.asJava).values().asScala.toSeq
>       }
> ```

Agree

pan3793 commented on client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:
```diff
@@ -115,12 +115,15 @@ class WorkerStatusTracker(
       val blacklistMsg = blacklist.asScala.map { case (worker, (status, time)) =>
         s"${worker.readableAddress()}   ${status.name()}   $time"
       }.mkString("\n")
+      val shuttingDownMsg = shuttingWorkers.asScala.map { _.readableAddress() }.mkString(";")
```
```suggestion
      val shuttingDownMsg = shuttingWorkers.asScala.map(_.readableAddress()).mkString("\n")
```

waitinfuture commented:
> 

We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the correct list.

waitinfuture commented:
Hi @pan3793 @RexXiong @zhongqiangczq , I've addressed the comments and refined the logic

pan3793 commented on client/src/main/scala/org/apache/celeborn/client/WorkerStatusTracker.scala:
```diff
@@ -53,10 +53,18 @@ class WorkerStatusTracker(
   }
 
-  def isWorkerAvailable(worker: WorkerInfo) = {
+  def workerAvailable(worker: WorkerInfo) = {
```
The function name `isXXX` is fine, it's consistent with others `verbXYZ`
```
isWorkerAvailable
isWorkerOfPartitionExcluded
```

waitinfuture replied:
I feel it a bit verbose..

waitinfuture replied:
done

pan3793 commented:
> We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.

Thanks for the explanation, it's worth to comment it in the code and adding it to the PR description.

RexXiong commented on client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:
```diff
@@ -587,6 +589,9 @@ private boolean revive(
               ClassTag$.MODULE$.apply(PbChangeLocationResponse.class));
       // per partitionKey only serve single PartitionLocation in Client Cache.
       StatusCode respStatus = Utils.toStatusCode(response.getStatus());
+      if (!response.getExcluded()) {
```
Why not remove newLocation worker from blacklist, And IMO client side blacklist would sometimes more timely then LifecycleManager, so it maybe not credible from LM.

waitinfuture replied:
The newLocation should not be in executor's excluded list because revive will exclude the oldPartition. Executor's excluded list will be added in LifecycleManager's excluded list, see ```ChangePartitionManager::handleRequestPartitions```. Currently, LM has a more complex and accurate strategy to mange its excluded list, so I think it's more accurate than executor side.

waitinfuture commented on client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:
```diff
@@ -226,9 +229,11 @@ class ChangePartitionManager(
+          req.context.reply(
+            StatusCode.SUCCESS,
+            Option(newLocation),
+            lifecycleManager.workerStatusTracker.excluded(req.oldPartition))))
```
done

pan3793 replied:
I see, `requests.toSeq...` could fix it.

waitinfuture commented:
> > We want a worker be either in the excluded list, or in the shutting down list, exclusively. If a worker is in excluded list, then LIfecycleManager will not trigger commit files when handle heartbeat response; on the other hand, if a worker is in the shutting down list, LifecycleManager will trigger commit files on it. So we must make it correct that a shutting down worker be in the shutting down list.
> 
> Thanks for the explanation, it's worth to comment it in the code and adding it to the PR description.

Updated the description. Since I reverted the modification to Master, and instead changes Worker's heartbeat, I don't put the comment there.

waitinfuture commented:
Hi @RexXiong @pan3793 , I've addressed the comments, please take a look again.

RexXiong reviewed (approved):
LGTM

pan3793 commented on common/src/main/proto/TransportMessages.proto:
```diff
@@ -220,6 +220,7 @@ message PbRevive {
 message PbChangeLocationResponse {
   int32 status = 1;
   PbPartitionLocation location = 2;
+  bool available = 3;
```
I suppose the 0.2 client just simply ignores it and still works well, right?

waitinfuture replied:
This PR is OK, but I do find compatibility exceptions when testing 0.2.1 client with main server. I will investigate and send new PRs to solve.

pan3793 commented on client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala:
```diff
@@ -226,9 +229,12 @@ class ChangePartitionManager(
-        requests.map(_.asScala.toList.foreach(_.context.reply(
-          StatusCode.SUCCESS,
-          Option(newLocation))))
+        requests.foreach(_.asScala.foreach { req =>
```
why this change is reverted?
